### PR TITLE
[RISCV] CodeGen of RVE and ilp32e/lp64e ABIs

### PR DESCRIPTION
**Pull request by wangpc-pp**
This commit includes the necessary changes to clang and LLVM to support
codegen of `RVE` and the `ilp32e`/`lp64e` ABIs.

The differences between `RVE` and `RVI` are:
* `RVE` reduces the integer register count to 16(x0-x16).
* The ABI should be `ilp32e` for 32 bits and `lp64e` for 64 bits.

`RVE` can be combined with all current standard extensions.

The central changes in ilp32e/lp64e ABI, compared to ilp32/lp64 are:
* Only 6 integer argument registers (rather than 8).
* Only 2 callee-saved registers (rather than 12).
* A Stack Alignment of 32bits (rather than 128bits).
* ilp32e isn't compatible with D ISA extension.

If `ilp32e` or `lp64` is used with an ISA that has any of the registers
x16-x31 and f0-f31, then these registers are considered temporaries.

To be compatible with the implementation of ilp32e in GCC, we don't use
aligned registers to pass variadic arguments and set stack alignment\
to 4-bytes for types with length of 2*XLEN.

FastCC is also supported on RVE, while GHC isn't since there is only one
avaiable register.

Differential Revision: https://reviews.llvm.org/D70401

### CODE REVIEW

**Comment by topperc** [comment]
What was the last bit of discussion on the phabricator review? I can no longer access it.

File: clang/lib/Driver/ToolChains/Arch/RISCV.cpp
Comment (topperc) [inline clang/lib/Driver/ToolChains/Arch/RISCV.cpp:297]: `equals_insensitive` like lip32e?

File: llvm/lib/Target/RISCV/MCTargetDesc/RISCVBaseInfo.cpp
Comment (topperc) [inline llvm/lib/Target/RISCV/MCTargetDesc/RISCVBaseInfo.cpp:74]: "must not" -> "cannot"

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:516]: Does this comment about 16 byte aligned need to be updated?

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:17191]:
```
Align StackAlign(XLenInBytes);
if (!EABI || XLen != 32)
  StackAlign = std::max(StackAlign, ArgFlags1.getNonZeroOrigAlign());
```

File: llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp:120]: `MCPhysReg`.  Registers aren't size_t.

File: llvm/test/CodeGen/RISCV/calling-conv-lp64e.ll
Comment (topperc) [inline llvm/test/CodeGen/RISCV/calling-conv-lp64e.ll:10]: This comment isn't accurate for this test.

File: llvm/test/CodeGen/RISCV/calling-conv-ilp32e.ll
Comment (topperc) [inline llvm/test/CodeGen/RISCV/calling-conv-ilp32e.ll:16]: This comment isn't accurate for this test.

**Comment by athei** [comment]
> What was the last bit of discussion on the phabricator review? I can no longer access it.

It was basically just waiting for final approval. The agreement was that it is good enough to be merged in as experimental even with unfinalized ABI.

I really hope we can get this in before end of this month to be able to use it in rustc.

**Comment by asb** [comment]
The conclusion from the previous review was this was OK to merge. I think I held it up by not responding to a ping (apologies). I've had another scan through and don't see a problem with merging this and considering it experimental once Craig's review comments are addressed.

For the psABI, there's a gap here I think as GCC and LLVM need to deviate from what is currently written. I think https://github.com/riscv-non-isa/riscv-elf-psabi-doc/pull/257 tried to partially address this. I wonder if we need a tracking issue in that repo to try to get this properly documented via that PR or otherwise (what do you think @jrtc27 @kito-cheng?).

**Comment by kito-cheng** [comment]
Hmmm, RISC-V ISA is growth after https://github.com/riscv-non-isa/riscv-elf-psabi-doc/pull/257 again, I mean...we have zfinx and zdinx, which is also valid combination with rv32e/rv64e, so we may need to revise ilp32e ABI again on the psABI side, but my intention is not to block this PR.

Anyway, I think we could track that on psABI repo, and we...already has an old issue to track that: https://github.com/riscv-non-isa/riscv-elf-psabi-doc/issues/269

**Comment by kito-cheng** [comment]
@wangpc-pp did you have interested on helping psABI side? it would be great if you can help since I suspect I don't have enough bandwidth to deal with that soon.

**Comment by wangpc-pp** [comment]
> @wangpc-pp did you have interested on helping psABI side? it would be great if you can help since I suspect I don't have enough bandwidth to deal with that soon.

Yes, I'm glad to. I think what we need to do is to fix some Zdinx issues? :-)

And, I think I have to explain the background here: 
I started to maintain this patch about two years ago when I was in T-head, since they have some old RV32E cores (AFAIK, SiFive has RV32E cores too?) and some customers needed the support of RV32E/ilp32e. Now, both T-head and SiFive didn't release RV32E cores any more, and the RVI doesn't put its focuses on RVE too I think. But I don't think there is no vendor that developed or developing their own RV32E implementations.
Now I am no longer working at T-head and my current work is not about RV32E, so I don't have environment to test it. But I think this patch can work just fine since it has been used in T-head's downstream for a long time (since LLVM 13, IIRC) and @zixuan-wu has reported several issues (already solved) before.
Apart from vendors' need, the request is mainly from Rust language community I think. IIUC, this is for the Rust implementation of SBI (https://github.com/rustsbi/rustsbi). They are using T-head E902 (rv32em?) and RV32E core from SiFive.

So, yeah, I will try my best to fix the ABI issue, but I think we still need the help of the RISC-V community. I will contact with T-head guys to see if they are interested in this.

**Comment by asb** [comment]
@nemanjai I'm curious if you have an interest / need to support RVE or not?

File: llvm/docs/RISCVUsage.rst
Comment (asb) [inline llvm/docs/RISCVUsage.rst:186]: Maybe this would be a good place to add "To be compatible with the implementation of ilp32e in GCC, we don't use aligned registers to pass variadic arguments and set stack alignment to 4-bytes for types with length of 2*XLEN."

Given this gap in the ABI, it would be good if we could document it somewhere outside of the commit message.

**Comment by koute** [comment]
> But I think this patch can work just fine [...] Apart from vendors' need, the request is mainly from Rust language community I think.

To put our 3 cents here, [we](https://github.com/paritytech) also need it. We are (well, it's mostly me right now since it's still a prototype) currently building [a virtual machine that is based on RV32E](https://github.com/koute/polkavm) (and in the future also RV64E), and, long story short, thanks to the magic of RV32E it achieves execution performance roughly the same as [wasmtime](https://github.com/bytecodealliance/wasmtime) (which is a state-of-art WASM VM) while compiling into native code over ~200 times faster and being orders of magnitude simpler (because you can mostly naively translate RV32E machine code 1-to-1 into native AMD64 code as the reduced number of registers makes this barely possible, and the code is still very fast).

As far as testing this patch, as a fairly decent test case I've used one of the previous versions to compile DOOM into RV32E and ran it under our VM (in fact, [you can play it here it you want](https://github.com/koute/polkavm/tree/master/examples/doom)), so the patch definitely works.

Anyway, people are underestimating RV32E thinking it's only for bottom tier microcontrollers, but this couldn't be further from the truth. It's not perfect, but it's also a really great VM bytecode; dare I say, I think it's a better WASM than WASM for non-Web uses.

**Comment by koute** [comment]
Side note: shouldn't we also update `compiler-rt/lib/builtins/riscv/{save, restore}.S`? E.g. with something like this:

```patch
diff --git a/compiler-rt/lib/builtins/riscv/restore.S b/compiler-rt/lib/builtins/riscv/restore.S
index 73f64a920d66..2e185ecae3f7 100644
--- a/compiler-rt/lib/builtins/riscv/restore.S
+++ b/compiler-rt/lib/builtins/riscv/restore.S
@@ -20,6 +20,46 @@
 
   .text
 
+#if __riscv_abi_rve == 1
+
+#if __riscv_xlen == 32
+  .globl  __riscv_restore_2
+  .type   __riscv_restore_2,@function
+  .globl  __riscv_restore_1
+  .type   __riscv_restore_1,@function
+  .globl  __riscv_restore_0
+  .type   __riscv_restore_0,@function
+__riscv_restore_1:
+__riscv_restore_0:
+  lw      s1,  4(sp)
+  lw      s0,  8(sp)
+  lw      ra,  12(sp)
+  addi    sp, sp, 16
+  ret
+#elif __riscv_xlen == 64
+  .globl  __riscv_restore_2
+  .type   __riscv_restore_2,@function
+__riscv_restore_2:
+  ld      s1,  8(sp)
+  addi    sp, sp, 16
+  // fallthrough into __riscv_restore_1/0
+
+  .globl  __riscv_restore_1
+  .type   __riscv_restore_1,@function
+  .globl  __riscv_restore_0
+  .type   __riscv_restore_0,@function
+__riscv_restore_1:
+__riscv_restore_0:
+  ld      s0,  0(sp)
+  ld      ra,  8(sp)
+  addi    sp, sp, 16
+  ret
+#else
+# error "xlen must be 32 or 64 for save-restore implementation
+#endif
+
+#else
+
 #if __riscv_xlen == 32
 
   .globl  __riscv_restore_12
@@ -164,3 +204,5 @@ __riscv_restore_0:
 #else
 # error "xlen must be 32 or 64 for save-restore implementation
 #endif
+
+#endif
diff --git a/compiler-rt/lib/builtins/riscv/save.S b/compiler-rt/lib/builtins/riscv/save.S
index 85501aeb4c2e..86b62a369cd0 100644
--- a/compiler-rt/lib/builtins/riscv/save.S
+++ b/compiler-rt/lib/builtins/riscv/save.S
@@ -16,6 +16,54 @@
 
   .text
 
+#if __riscv_abi_rve == 1
+
+#if __riscv_xlen == 32
+  .globl  __riscv_save_2
+  .type   __riscv_save_2,@function
+  .globl  __riscv_save_1
+  .type   __riscv_save_1,@function
+  .globl  __riscv_save_0
+  .type   __riscv_save_0,@function
+__riscv_save_3:
+__riscv_save_2:
+__riscv_save_1:
+__riscv_save_0:
+  addi    sp, sp, -16
+  sw      s1,  4(sp)
+  sw      s0,  8(sp)
+  sw      ra,  12(sp)
+  jr      t0
+#elif __riscv_xlen == 64
+  .globl  __riscv_save_2
+  .type   __riscv_save_2,@function
+__riscv_save_3:
+__riscv_save_2:
+  addi   sp, sp, -112
+  li     t1, 80
+.Lriscv_save_3_2:
+  sd     s1, 88(sp)
+  sd     s0, 96(sp)
+  sd     ra, 104(sp)
+  add    sp, sp, t1
+  jr     t0
+
+  .globl  __riscv_save_1
+  .type   __riscv_save_1,@function
+  .globl  __riscv_save_0
+  .type   __riscv_save_0,@function
+__riscv_save_1:
+__riscv_save_0:
+  addi   sp, sp, -16
+  sd     s0, 0(sp)
+  sd     ra, 8(sp)
+  jr     t0
+#else
+# error "xlen must be 32 or 64 for save-restore implementation
+#endif
+
+#else
+
 #if __riscv_xlen == 32
 
   .globl  __riscv_save_12
@@ -184,3 +232,5 @@ __riscv_save_0:
 #else
 # error "xlen must be 32 or 64 for save-restore implementation
 #endif
+
+#endif
```

(I don't remember why exactly since I did it a long time ago, but for some reason I do have this patch in my LLVM fork, so it probably was necessary for *something*.)

**Comment by wangpc-pp** [comment]
> Side note: shouldn't we also update `compiler-rt/lib/builtins/riscv/{save, restore}.S`? E.g. with something like this:
>  [...]
> (I don't remember why exactly since I did it a long time ago, but for some reason I do have this patch in my LLVM fork, so it probably was necessary for _something_.)

Oh, it may be about stack alignment issue for libcalls. I may have fixed it.
As for your diffs, it seems that you only handle the `__riscv_save/restore_[2|1|0]`, which is incomplete. And the code is not different with non-rve cases?

**Comment by koute** [comment]
> As for your diffs, it seems that you only handle the `__riscv_save/restore_[2|1|0]`, which is incomplete. And the code is not different with non-rve cases?

Yes, I mostly copy-pasted the existing code and removed all of the code dealing with registers not available on RV32E, so having only `__riscv_save/restore_[2|1|0]` is intended I suppose because there are only this many saved registers on RV32E. (There's probably a better way of doing it, and it looks like I screwed up the RV64E part of the patch.)

If I remember correctly I think I did this because otherwise compiling `compiler-rt` was not possible for RV32E and the compilation was spewing out errors about the unavailable registers? But I need to check this again once I finish porting this newest version of the patch to the most recent version of Rust.

**Comment by wangpc-pp** [comment]
> > As for your diffs, it seems that you only handle the `__riscv_save/restore_[2|1|0]`, which is incomplete. And the code is not different with non-rve cases?
> 
> Yes, I mostly copy-pasted the existing code and removed all of the code dealing with registers not available on RV32E, so having only `__riscv_save/restore_[2|1|0]` is intended I suppose because there are only this many saved registers on RV32E. (There's probably a better way of doing it, and it looks like I screwed up the RV64E part of the patch.)
> 
> If I remember correctly I think I did this because otherwise compiling `compiler-rt` was not possible for RV32E and the compilation was spewing out errors about the unavailable registers? But I need to check this again once I finish porting this newest version of the patch to the most recent version of Rust.

Oh, I see. My previous comment was wrong.
I just checked GCC implementation, we do need to handle RVE cases (but GCC still lacks of RV64E handling). LLVM hasn't handled this mainly because we haven't support RVE now, I think.
Can you fire a PR for these changes? I think we should support it in compiler-rt once we have merge this PR.

**Comment by koute** [comment]
> Can you fire a PR for these changes? I think we should support it in compiler-rt once we have merged this PR.

Once this PR is merged, sure, I can make a PR.

**Comment by nemanjai** [comment]
> @nemanjai I'm curious if you have an interest / need to support RVE or not?

I most certainly do. Thank you for alerting me to this PR.

**Comment by asb** [comment]
The RISCVUsage change looks good to me, and it's fantastic that @nemanjai might be able to help push this forward further.

I think all the review comments were addressed, so I'd be happy to merge if @topperc is happy his comments were resolved. Though also fine to wait longer if @nemanjai you wanted to review this as well.

**Review by nemanjai — COMMENTED** [review]
It is not my intent to hold up approval of this patch. In addition to the minor comments I added, I plan to do some local testing but even if the testing reveals issues, they can be fixed on subsequent commits.

File: llvm/lib/Target/RISCV/RISCVFeatures.td
Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVFeatures.td:26]: For `RVE`, we will have some command lines like `-target-feature -i` in clang driver, which will cause error as we don't have `i` extension in backend. This fixes this error.

File: llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp
Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp:91]: This should be a rebase mistake.

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp
Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:1006]: The psABI says:
> If used with an ISA that has any of the registers x16-x31 and f0-f31, then these registers are considered temporaries.

So I think we should.

File: clang/lib/Basic/Targets/RISCV.cpp
Comment (tclin914) [inline clang/lib/Basic/Targets/RISCV.cpp:400]: Does it also need to check ilp32e isn't compatible with zcmp?

Comment (wangpc-pp) [reply clang/lib/Basic/Targets/RISCV.cpp:400]: Yes, I think we should!
(Sigh...even RISC-V International has forgotten RVE, as many new ISA extensions don't even think about impacts on RVE 😞).

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp
Comment (nemanjai) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:1006]: Sounds good. Maybe just a little comment here to demonstrate to the user that this isn't an omission but a conscious decision.

**Review by nemanjai — APPROVED** [review]
My comments have been addressed, so this LGTM. I'll of course defer to @asb and @topperc for final approval.

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp
Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:1007]: Though it's nearly impossible to have such configuration in real application, I saved x15-x16 in https://github.com/llvm/llvm-project/pull/76777/commits/20ffba38548c823d6ef286bcad63385087438d57.

**Review by asb — APPROVED** [review]
LGTM. Thanks for your persistence on this!

**Comment by wangpc-pp** [comment]
It's not easy to push this forward and thanks a lot!
I'll merge this soon. If there are some issues (hope not 😄), follow-up patches will fix them.

**Comment by koute** [comment]
As a third-party bystander I just want to say, thank you to everyone involved! And especially @wangpc-pp, a huge thank you for pushing this through!

**Pull request merged**
